content/blog/2023-04-01-key-expressions-1.md

```diff
@@ -0,0 +1,147 @@
+
+# The KE Tree: a Zenoh flavoured data structure.
+
+One thing you might have noticed is that with all these slashes, KEs definitely take after paths. One other aspect they take from paths is their intrisic hierarchical nature: like most good address spaces, KEs are hierarchical.
+
+And what data-structure rimes with hierarchy? The section title spoiled it, it's the _tree_. But the KeTree isn't just any old tree: it's a tree that's made to help you treat KEs as the sets they represent, complete with _intersection_ and _inclusion_ comparisons.
+
+## Wait, what does it mean for KEs to _intersect_ and _include_?
+
+Glad you asked, inner monologue of my future reader, forcing you to read this question in your mind was definitely helpful!
+
+As aluded to repeatedly in this post, KEs define sets of keys, and operations in Zenoh are addressed by KE, so they affect sets of things. In fact, there are four ways an operation on a KE can affect data associated with a given KE:
+1. If the KEs define disjoint sets (where there doesn't exist a key that belongs to both), they just don't interract.
+2. If the operation KE and the existing data's KE define sets that have keys in common, like `a/*` and `*/b` both contain the `a/b` key, they are said to _intersect_. That means the operation will affect at least a subset of the existing data. Intersection is always symmetric: A intersects with B implies that B intersects with A.
+3. If the operation KE's set contains all keys defined by the existing data's KE, the operation KE is said to _include_ the existing data KE. This means that the operation will affect _all_ of the existing data. Note that inclusion is generally asymmetric, but that "A includes B" or "B includes A" implies that "A and B intersect".
+4. If the operation KE and data KE define the same set, they are equal. This is the only situation where inclusion is symmetrical, and thanks to our previously discussed [3 steps program](#specifying-the-kel), this is equivalent to string equality.
+
+Intersection is often the most important comparison, since it means that two KEs have to interract in some way, because they share a region of interest. This is the criterion that Zenoh uses to route samples to subscribers, and you'll likely use this criterion too, applied to your business logic, when working with queryables.
+
+Inclusion is a bit more "optimizy": if some writes to A that includes B and C, the records for B and C may be erased, since A has now taken over both of their regions on interest.
+
+Equality and disjunction are generally not very useful: equality because inclusion is generally sufficient for most optimizations, and disjunction because it usually just means that two things do not care about each other, which we in turn don't care about.
```

Comment:
I think this sentence would be better as a footnote

Reply:
I don't really disagree, considering it's not key information, it just feels more natural to me to include all 4 relations in the same flow rather than interrupting it with an indirection. Plus, I wouldn't really know where to put the footnote marker x)